### ISSUE TITLE
netcup dns validation issue (no TXT record written)

### ISSUE THREAD

**Issue by LivingWithHippos** [linuxserver/docker-swag#44]
[linuxserverurl]: https://linuxserver.io
[![linuxserver.io](https://raw.githubusercontent.com/linuxserver/docker-templates/master/linuxserver.io/img/linuxserver_medium.png)][linuxserverurl]

------------------------------

## Expected Behavior

A wildcard SSL certificate is issued for my domain on netcup

## Current Behavior

No TXT record is added to my domain

## Steps to Reproduce

1. prepare the docker-compose file
2. create the `dns-conf/netcup.ini` file
3. run docker-compose up -d

## Environment
**OS:** 
ubuntu 20
Linux 5.4.0-56-generic #62-Ubuntu SMP Mon Nov 23 19:20:19 UTC 2020 x86_64 x86_64 x86_64 GNU/Linu
**CPU architecture:** x86_64
**How docker service was installed:**
official docker repo

## Command used to create docker container (run/create/compose/screenshot)

`docker-compose.yml`

```yaml
---
version: "3.5"
services:
  swag:
    image: ghcr.io/linuxserver/swag
    container_name: swag
    cap_add:
      - NET_ADMIN
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=Europe/London
      - URL=domain.tld
      - SUBDOMAINS=wildcard
      - VALIDATION=dns
      - DNSPLUGIN=netcup
      - EMAIL= mail@gmail.com
      - ONLY_SUBDOMAINS=false #optional
      #- EXTRA_DOMAINS= secondomain.com,www.secondomain.it,*.thirddomain.com
      - STAGING=false #optional
      #- MAXMINDDB_LICENSE_KEY= #optional
    volumes:
      - /docker/swag/config:/config
    ports:
      - 443:443
      - 80:80 #optional
    restart: unless-stopped
    networks:
      - letsencrypt_default

networks:
  letsencrypt_default:
    external: true
```

## Docker logs

`docker logs swag`

```
...
-------------------------------------
GID/UID
-------------------------------------

User uid:    1000
User gid:    1000
-------------------------------------

[cont-init.d] 10-adduser: exited 0.
[cont-init.d] 20-config: executing...
[cont-init.d] 20-config: exited 0.
[cont-init.d] 30-keygen: executing...
generating self-signed keys in /config/keys, you can replace these with your own keys if required
Generating a RSA private key
....+++++
.........................................+++++
writing new private key to '/config/keys/cert.key'
-----
[cont-init.d] 30-keygen: exited 0.
[cont-init.d] 50-config: executing...
Variables set:
PUID=1000
PGID=1000
TZ=Europe/London
URL=domain.tld
SUBDOMAINS=wildcard
EXTRA_DOMAINS=
ONLY_SUBDOMAINS=false
VALIDATION=dns
DNSPLUGIN=netcup
EMAIL= mail@gmailcom
STAGING=false

Created donoteditthisfile.conf
SUBDOMAINS entered, processing
Wildcard cert for domain.tld will be requested
E-mail address entered: mail@gmailcom
dns validation via netcup plugin is selected
Generating new certificate
Use of --manual-public-ip-logging-ok is deprecated.
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Plugins selected: Authenticator dns-netcup, Installer None
Account registered.
Requesting a certificate for *.domain.tld and domain.tld
Performing the following challenges:
dns-01 challenge for domain.tld
dns-01 challenge for domain.tld
Unsafe permissions on credentials configuration file: /config/dns-conf/netcup.ini
Waiting 10 seconds for DNS changes to propagate
Waiting for verification...
Challenge failed for domain domain.tld
Challenge failed for domain domain.tld
dns-01 challenge for domain.tld
dns-01 challenge for domain.tld
Cleaning up challenges
Some challenges have failed.
IMPORTANT NOTES:
 - The following errors were reported by the server:

   Domain: domain.tld
   Type:   unauthorized
   Detail: No TXT record found at _acme-challenge.domain.tld

   Domain: domain.tld
   Type:   unauthorized
   Detail: No TXT record found at _acme-challenge.domain.tld

   To fix these errors, please make sure that your domain name was
   entered correctly and the DNS A/AAAA record(s) for that domain
   contain(s) the right IP address.
ERROR: Cert does not exist! Please see the validation error above. Make sure you entered correct credentials into the /config/dns-conf/netcup.ini file.
```

I tracked down the issue in `/var/log/letsencrypt/letsencrypt.log`

```
2020-12-06 17:43:52,149:DEBUG:acme.client:Storing nonce: 0004lVrIOjytUY9S1by5XIaN5if1tMesHvBHUZZqI7c3M
2020-12-06 17:43:52,149:INFO:certbot._internal.auth_handler:Performing the following challenges:
2020-12-06 17:43:52,149:INFO:certbot._internal.auth_handler:dns-01 challenge for domain.tld
2020-12-06 17:43:52,150:INFO:certbot._internal.auth_handler:dns-01 challenge for domain.tld
2020-12-06 17:43:52,150:WARNING:certbot.plugins.dns_common:Unsafe permissions on credentials configuration file: /config/dns-conf/netcup.ini
2020-12-06 17:43:52,152:DEBUG:lexicon.providers.netcup:login({})
2020-12-06 17:43:52,153:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): ccp.netcup.net:443
2020-12-06 17:43:52,198:DEBUG:urllib3.connectionpool:https://ccp.netcup.net:443 "POST /run/webservice/servers/endpoint.php?JSON HTTP/1.1" 200 228
2020-12-06 17:43:52,200:DEBUG:lexicon.providers.netcup:infoDnsZone({'domainname': 'domain.tld'})
2020-12-06 17:43:52,201:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): ccp.netcup.net:443
2020-12-06 17:43:52,259:DEBUG:urllib3.connectionpool:https://ccp.netcup.net:443 "POST /run/webservice/servers/endpoint.php?JSON HTTP/1.1" 200 238
2020-12-06 17:43:52,261:DEBUG:lexicon.providers.netcup:infoDnsRecords({'domainname': 'domain.tld'})
2020-12-06 17:43:52,262:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): ccp.netcup.net:443
2020-12-06 17:43:52,327:DEBUG:urllib3.connectionpool:https://ccp.netcup.net:443 "POST /run/webservice/servers/endpoint.php?JSON HTTP/1.1" 200 319
2020-12-06 17:43:52,329:DEBUG:lexicon.providers.netcup:list_records: []
2020-12-06 17:43:52,329:DEBUG:lexicon.providers.netcup:updateDnsRecords({'domainname': 'domain.tld', 'dnsrecordset': {'dnsrecords': [{'type': 'TXT', 'hostname': '_acme-challenge', 'destination': 'sp412mXHfTR1DyJ5EAp7AuhxHEgDKk2JxgePBy0BRMZ'}]}})
2020-12-06 17:43:52,330:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): ccp.netcup.net:443
2020-12-06 17:43:54,445:DEBUG:urllib3.connectionpool:https://ccp.netcup.net:443 "POST /run/webservice/servers/endpoint.php?JSON HTTP/1.1" 200 419
2020-12-06 17:43:54,448:DEBUG:lexicon.providers.netcup:create_record: True
2020-12-06 17:43:54,450:DEBUG:lexicon.providers.netcup:login({})
2020-12-06 17:43:54,451:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): ccp.netcup.net:443
2020-12-06 17:43:54,489:DEBUG:urllib3.connectionpool:https://ccp.netcup.net:443 "POST /run/webservice/servers/endpoint.php?JSON HTTP/1.1" 200 229
2020-12-06 17:43:54,491:DEBUG:lexicon.providers.netcup:infoDnsZone({'domainname': 'domain.tld'})
2020-12-06 17:43:54,492:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): ccp.netcup.net:443
2020-12-06 17:43:54,551:DEBUG:urllib3.connectionpool:https://ccp.netcup.net:443 "POST /run/webservice/servers/endpoint.php?JSON HTTP/1.1" 200 240
2020-12-06 17:43:54,553:DEBUG:lexicon.providers.netcup:infoDnsRecords({'domainname': 'domain.tld'})
2020-12-06 17:43:54,555:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): ccp.netcup.net:443
2020-12-06 17:43:54,613:DEBUG:urllib3.connectionpool:https://ccp.netcup.net:443 "POST /run/webservice/servers/endpoint.php?JSON HTTP/1.1" 200 394
2020-12-06 17:43:54,615:DEBUG:lexicon.providers.netcup:list_records: []
2020-12-06 17:43:54,616:DEBUG:lexicon.providers.netcup:updateDnsRecords({'domainname': 'domain.tld', 'dnsrecordset': {'dnsrecords': [{'type': 'TXT', 'hostname': '_acme-challenge', 'destination': 'SPdEcslSAc-uSMufAUUOIqHup5vp7L8WOrTZfGAT6PE'}]}})
2020-12-06 17:43:54,617:DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): ccp.netcup.net:443
2020-12-06 17:43:56,679:DEBUG:urllib3.connectionpool:https://ccp.netcup.net:443 "POST /run/webservice/servers/endpoint.php?JSON HTTP/1.1" 200 464
2020-12-06 17:43:56,681:DEBUG:lexicon.providers.netcup:create_record: True
2020-12-06 17:43:56,681:INFO:certbot.plugins.dns_common:Waiting 10 seconds for DNS changes to propagate
```

So the netcup login works, in fact I even have the API log from their side with a "correct" voice

Action | status | Feedback short | Domain name / Handle Id | API Key
-- | -- | -- | -- | --
updateDnsRecords | Successfully | DNS records successful updated | domain.tld | netcup.ini api key

continuing the log the dns record are checked and they fail.

```
2020-12-06 17:44:08,621:DEBUG:acme.client:Received response:
HTTP 200
Server: nginx
Date: Sun, 06 Dec 2020 17:44:08 GMT
Content-Type: application/json
Content-Length: 556
Connection: keep-alive
Boulder-Requester: 105323570
Cache-Control: public, max-age=0, no-cache
Link: <https://acme-v02.api.letsencrypt.org/directory>;rel="index"
Replay-Nonce: 000379E_YezNUePmp4hplBBraSUNLwm2ltX96PdXkKqyEjp
X-Frame-Options: DENY
Strict-Transport-Security: max-age=604800

{
  "identifier": {
    "type": "dns",
    "value": "domain.tld"
  },
  "status": "invalid",
  "expires": "2020-12-13T17:43:51Z",
  "challenges": [
    {
      "type": "dns-01",
      "status": "invalid",
      "error": {
        "type": "urn:ietf:params:acme:error:unauthorized",
        "detail": "No TXT record found at _acme-challenge.domain.tld",
        "status": 403
      },
      "url": "https://acme-v02.api.letsencrypt.org/acme/chall-v3/9118012780/OUnXUA",
      "token": "9D0OB3k2ri0JWeAZ1TAwOpxyKfW48Jdyj6xTtA2weRI"
    }
  ],
  "wildcard": true
}
```

checking from netcup side the dns record the _acme-challenge TXT is missing.

to recap:
- the netcup.ini seems to be correct (from the container and netcup logs)
- the domains A record should be okay (I had single https certificates on some subdomains)

I don' t know I think it' s an API issue, either on netcup side (why is there a " correct" voice on updateDNSrecord if there are no updates?) or the netcup plugin.

As a workaround do you think I could quickly manually add the record to my netcup dns by quickly grepping the TXT value from the logs?

**Comment by aptalca** [comment]
Netcup support was PRed by a community member. Bone if the team members even have the capability to test it.

This should probably be reported to the netcup plugin author: https://github.com/coldfix/certbot-dns-netcup

To answer the last question, no, this image does not support manual txt record creation as automation is required for renewals as well.

**Comment by LivingWithHippos** [comment]
Thanks, I'll check there.

As for my question what I meant is, since the image waits 10 seconds before checking the TXT record, if I sshed into the container, get the value from the file and manually add it to my dns records all ni 10 seconds, do you think it should work?

**Comment by aptalca** [comment]
It's a random string generated for each txt record, each time, not sure if certbot outputs that anywhere. You can try, I guess.

**Comment by eggplanting** [comment]
Same issue here, adjusting the value of environment variable "PROPAGATION" didn't help either. I increased the default value from 10 seconds up to 120 seconds, but the issue remains.

Only having this issue when a wildcard cert is issued via netcup plugin, Subdomains are working.

Would be great to see an update on this.

**Comment by LivingWithHippos** [comment]
I switched to another solution but I think this is the right fix, maybe it should be noted in the readme
